projects/npm-tools/packages/npm-scripts/src/scripts/build.js

```diff
@@ -82,6 +83,13 @@ module.exports = async function (...args) {
 		'liferay-npm-scripts: `build`'
 	);
 
+	const hasJS =
+		fs.existsSync(BUILD_CONFIG.input) &&
+		!!expandGlobs(
+			['*.js', '*.jsx', '*.ts', '*.tsx'],
+			['build', 'classes', 'node_modules']
```

Comment:
Should you be anchoring this search in `BUILD_CONFIG.input`? Just in case there is some module out there that has a top-level `.js` file (eg. a config file?)

Kind of sad that we have to do this, and that the two lists here have to be so hard-coded (well, really, the hard-coding of the second list is the one that is a bit sad).

I wonder how feasible it is to fix this "upstream". For example, maybe the bundler could be taught to handle the no-JS case gracefully/silently instead of:

```
Error: Loader 'undefined' failed: Cannot determine web context path for the project: either specify it in .npmbundlerrc or bnd.bnd
```

@izaera: thoughts?

About Babel, presumably that's the one that is printing this:

```
src/main/resources/META-INF/resources does not exist
```

Here again I wonder exactly where that error is coming from and whether we could just teach it to be graceful. 🤔

Reply:
The error about the web context is thrown from `liferay-npm-bundler-loader-css-loader` when it tries to generate the `.js` module for a `.scss` file.

I would say there's no way to fix that upstream since you are calling the bundler in a project that is not meant to call the bundler so, even if we ignore that, we may get into lots of unpredictable problems. Also, what's the point of running a tool that, in the best case, it's going to do nothing? :sweat_smile: 

Maybe we can detect this by looking for a `bnd.bnd` file for example, in addition to `.js` files. That would fix the bundler problem.

Regarding babel, I suppose we only need to check if `src/main/resources/META-INF/resources` exists. Given that it is given as configuration to `npm-scripts` we don't have to hardcode it, I think (I think that is inside `BUlLD_CONFIG.input`...).

Reply:
Yes, it is:
![image](https://user-images.githubusercontent.com/3273630/102328778-cef17e80-3f87-11eb-805a-24605e70bdb4.png)

Reply:
> Maybe we can detect this by looking for a `bnd.bnd` file for example, in addition to `.js` files. That would fix the bundler problem.
>
> ...
>
> I suppose we only need to check if `src/main/resources/META-INF/resources` exists.

This sounds better than recursively traversing an entire directory hierarchy, if looking at the `bnd.bnd` alone is enough... basically just looking for a way to do the check in `O(1)` instead of `O(n = size of module)`.

Reply:
Looks like babel is happy enough with just the check if the `BUlLD_CONFIG.input` path exists.


For the bundler...

> looking for a bnd.bnd file for example, in addition to .js files.

@wincent sounds like we would still need to check for the js files _in addition_. But I think if we anchor the search for js files with `BUlLD_CONFIG.input` it should cut costs considerably.

I'll investigate this a bit more and see if we can reduce the overhead of traversing the file structure multiple times (once for the check, additionally for bundler & babel).

Reply:
> But I think if we anchor the search for js files with `BUlLD_CONFIG.input` it should cut costs considerably.

Yeah, will avoid any potentially nasty nested `node_modules` subtrees that might be hiding in the module root.